### ISSUE TITLE
EclipseLink DB2Platform no longer binds Literals as Parameters (DB2Platform.java#L74 => "this.shouldBindLiterals = false;")

### ISSUE THREAD

**Issue by hubmat00** [eclipse-ee4j/eclipselink#866]
EclipseLink DB2Platform no longer binds Literals as Parameters (DB2Platform.java#L74 => this.shouldBindLiterals = false;)

We are using EclipseLink packaged within Websphere Liberty Profile (WLP) version 20.0.0.x in combination with DB2. WLP 20.0.0.x packages EclipseLink in the version 2.7.6.

While updating from WLP version 19.0.0.x to 20.0.0.x we experience problems with the new provided EclipseLink version 2.7.6. As changed in the DB2Platform with version 2.7.6 (https://github.com/eclipse-ee4j/eclipselink/blob/master/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/platform/database/DB2Platform.java#L74) literals get no longer bound as parameters in the query when using Criteria Builder, literals are bound inline in the SQL-statements. **_This behaviour leads to changed SQL-statements and restricts DB2 to NO longer use prepared statements with parameters._**

Changes that force DB2Platform to "this.shouldBindLiterals = false" were implemented with:
* EclipseLink Issue: https://github.com/eclipse-ee4j/eclipselink/issues/492
* EclipseLink Commit: https://github.com/eclipse-ee4j/eclipselink/commit/418a2e5506c36854e72a5fd5662cdff4ac1cb80e
* IBM related issue:
https://www.ibm.com/support/pages/apar/PH06475

There exists another IBM related Issue (https://www.ibm.com/support/pages/apar/PH14747), but I am not sure if this IBM issue relates to my issue.

In general and independent of known workarounds, **we ask for a global solution that allows to set the property "shouldBindLiterals" via system properties** (similar to "eclipselink.jdbc.force-bind-parameters" and "eclipselink.jdbc.bind-parameters" => https://github.com/eclipse-ee4j/eclipselink/blob/master/jpa/org.eclipse.persistence.jpa/src/main/java/org/eclipse/persistence/internal/jpa/EntityManagerSetupImpl.java#L2828 and L#2833). We are also open for other global solutions or fixes that allow us to bind literals as parameters!

### For WLP Applications with DB2 persistency we found some workarounds:
1) **Parameter-Expressions:** setting parameters as Parameter-Expressions => every query based on Criteria API needs to be adapted, not practicable for large applications with thousands of queries
```java
int value = 100;
CriteriaBuilder cb = em.getCriteriaBuilder();
CriteriaQuery<Entity> cq = cb.createQuery(Entity.class);
Root<Entity> root = cq.from(Entity.class);
ParameterExpression<Integer> parameter = cb.parameter(Integer.class);
cq.where(cb.equal(root.get(Entity_.intAttribute), parameter));
return em.createQuery(cq).setParameter(parameter, value).getResultList();
```
2) **Property in persistence.xml:** settings the "eclipselink.target-database-properties" for each PU in each persistence.xml to "shouldBindLiterals=true" => every persistence.xml needs to be changed, not practicable for large applications with hundreds of persistence.xml's
```
	<persistence-unit name="pu-name" transaction-type="JTA">
		<jta-data-source>jdbc/name</jta-data-source>

		...

		<exclude-unlisted-classes>true</exclude-unlisted-classes>
		<shared-cache-mode>NONE</shared-cache-mode>

		<properties>
			<property name="eclipselink.target-database-properties" value="shouldBindLiterals=true"/>
		</properties>

	</persistence-unit>
```
3) **WLP Custom Persistence Provider:** defining a customer persistence provider that adds the property-value "shouldBindLiterals=true" to the property "eclipselink.target-database-properties" of the proterties-map that is used to bootstrap the EntityManager
server.xml
```
...
<jpa defaultPersistenceProvider="com.provider.CustomPersistenceProvider"/>
...
```
CustomPersistenceProvider.java
```
import java.util.HashMap;
import java.util.Map;

import javax.persistence.EntityManagerFactory;
import javax.persistence.spi.PersistenceUnitInfo;

import org.eclipse.persistence.jpa.PersistenceProvider;

public class CustomPersistenceProvider
		extends PersistenceProvider {

	@Override
	public EntityManagerFactory createContainerEntityManagerFactory(PersistenceUnitInfo info, Map properties) {
		Map<String, Object> extendedProperties = new HashMap(properties);
		extendedProperties.put("eclipselink.target-database-properties", "shouldBindLiterals=true");
		return super.createContainerEntityManagerFactory(info, extendedProperties);
	}
}
```

All this known workarounds lead to extensive code canges and are not practicable for large application scopes.

Do you know other global and practicable workarounds for WLP applications with DB2 persistency that force EclipseLink in combination with Criteria API to bind literarls as parameters?

**Comment by benmoeARZ** [comment]
Same Problem here...
We would also need a global solution.

Option 1 is not really applicable because in a real big application you have to rewrite (and retest) all queries.

Option 2 is also a lot of work. You have to set it in every persistence.xml in your application, including persistence.xmls that comes to you in e.g. libs via dependency-management.

Option 3: I really tried that, but in our application that lead to really bad classloading problems.

So please guys help us :/

**Comment by ch4rl353y** [comment]
We're currently facing the same issue ...

**Comment by sojournercntl** [comment]
Unfortunately we have the same problem. We would be very thankful for help!

Thanks in advance 👍

**Comment by ch4rl353y** [comment]
any news on this issue?

**Comment by dazey3** [comment]
Hello!
I was the one who made the literal change, for DB2, in EclipseLink with Bug #542491 (#492). I was only recently made aware of this issue/complaint and I would like to start an investigation. 

---
>This behavior leads to changed SQL-statements
This is correct. The SQL statements themselves do change. I will summarize this change with an example:

Consider the following Criteria Builder use case:
```java
    int intValue = 100;
    ...
    Predicate intPredicate = criteriabuilder.equal(root.get(SimpleEntity_.intAttribute), value);
    criteriaquery.where(intPredicate);
    ...
    TypedQuery<SimpleEntity> query = em.createQuery(criteriaquery);
    List<SimpleEntity> result = query.getResultList();
```

**Previously**, before #492, EclipseLink would create a java.sql.PreparedStatement (via a call to `java.sql.Connection.prepareStatement(String sql)`) and execute the following SQL (via a call to `java.sql.PreparedStatement.executeQuery()`):
```
[EL Fine]: sql: ... SELECT ID, INTATTRIBUTE FROM SIMPLEENTITY WHERE (INTATTRIBUTE = ?)
	bind => [100]
```

**Now**, after #492, EclipseLink creates a java.sql.PreparedStatement (via a call to `java.sql.Connection.prepareStatement(String sql)`) and executes the following SQL (via a call to `java.sql.PreparedStatement.executeQuery()`):
```
[EL Fine]: sql: ... SELECT ID, INTATTRIBUTE, STRATTRIBUTE FROM SIMPLEENTITY WHERE (INTATTRIBUTE = 100)
```

The only difference I can observe in EclipseLink behavior is that prior to the #492 change, the java.sql.PreparedStatement was created with a parameter marker and the following call was made to set the Integer value on the PreparedStatement: `java.sql.PreparedStatement.setInt(int parameterIndex, int x)`
After the #492 change, this call to set the Integer parameter never occurred, because the literal value was not part of the query, not a parameter.

---
> restricts DB2 to NO longer use prepared statements with parameters

I'm not sure I understand this point. It is correct that with the #492 change, EclipseLink no longer will use parameter markers for literal values for DB2. However, it is not correct that EclipseLink will no longer use prepared statements with parameters. 
As demonstrated in the example by @hubmat00 , if you declare the value as a Criteria Builder parameter, then EclipseLink will use prepared statements with parameters, if that is the behavior the user wishes to obtain. Consider the Criteria Builder use case from earlier with the following change:
```java
    int intValue = 100;
    ...
    // Addition of ParameterExpression to declare the value as a query parameter
    ParameterExpression<Integer> intParameter = criteriabuilder.parameter(Integer.class);
    Predicate intPredicate = criteriabuilder.equal(root.get(SimpleEntity_.intAttribute), intParameter);
    criteriaquery.where(intPredicate);
    ...
    TypedQuery<SimpleEntity> query = em.createQuery(criteriaquery);
    // Setting the parameter value to be used
    query.setParameter(intParameter, intValue);
    List<SimpleEntity> result = query.getResultList();
```
---

> In general and independent of known workarounds, we ask for a global solution that allows to set the property "shouldBindLiterals" via system properties (similar to "eclipselink.jdbc.force-bind-parameters" and "eclipselink.jdbc.bind-parameters" ... We are also open for other global solutions or fixes that allow us to bind literals as parameters!

It is correct that the `shouldBindLiterals` property is the correct place to revert this change if that is what is required. However, I am not sure what "global" solution is being asked for. The `eclipselink.jdbc.force-bind-parameters` and `eclipselink.jdbc.bind-parameters` are, to my knowledge, not "global solutions" as they also require being set as persistence properties. Correct me if I am wrong, but the complaint seems to be that setting `<property name="eclipselink.target-database-properties" value="shouldBindLiterals=true"/>` requires too many changes if a sufficient number of persistence units exist. I can certainly agree with that, but I'm not sure what solution is being asked for here and I would appreciate more suggestions of what would be sufficient.

One idea may be to use a JVM property. Currently, EclipseLink does not process JVM properties for persistence properties. However, other persistence providers like OpenJPA do. It is feasible that an enhancement to EclipseLink could be made to allow this JVM properties behavior that OpenJPA uses. However, that is a bit of a separate task and should be addressed independently of this specific issue (in my opinion).

**Comment by hubmat00** [comment]
@dazey3

thanks for your well-elaborated answer to my request!

> Overall, I would like to know the impact of this change outside of "the SQL changed". Were there performance impacts? Incorrect query results? Database failures?

The changed SQL statements in our case lead to two implications:

## 1. DB Performance: limitation of reusing prepared statements from the DB's statement cache
having statements with literal values being part of the query leads to the fact that the reusage of this particular prepared statements from DB's statements cache is NO longer possible as the literal value within the query changes. This slows down query execution by increasing the parsing and query planning overhead of each execution.

Consider the following query as example:

`[EL Fine]: sql: ... SELECT ID, INTATTRIBUTE FROM SIMPLEENTITY WHERE (INTATTRIBUTE = 100)`

The value `100` is part of the query. While repeating this query with different values for `INTATTRIBUTE` (e.g. `200`, `300` or `400`), the database has to parse every single SQL-Query and to create query plans accordingly rather than reusing cached statements from the DB.

Now consider the following 'optimized' query:

```
[EL Fine]: sql: ... SELECT ID, INTATTRIBUTE FROM SIMPLEENTITY WHERE (INTATTRIBUTE = ?)
	bind => [100]
```
The value `100` is no longer part of the query as it is bound as parameter. While repeating this 'optimized' query with different values for `INTATTRIBUTE` (e.g. `200`, `300` or `400`), the database optimizes the query execution by reusing the cached queries from the DB statement cache for different values for `INTATTRIBUTE`. 

The reusage of prepared statements speeds up query execution by decreasing the parsing and query planning overhead of each execution. However, this reusage of queries can only be achieved when binding literal values as parameters. The changes of https://github.com/eclipse-ee4j/eclipselink/issues/492 lead to changes in the base behaviour of how such literals are bound to queries.

**To summarize:**
 * Where is the performance impacted? => the performance is impacted by the usage of prepared statements from the db statement cache
 * Incorrect query results? => no incorrect query results
 * Database failures? => no failures

## 2. Application failures based on missing parameters
**These application failures based on missing parameters is NO GENERAL issue!** This issue is triggered by a third-party JDBC-driver. This driver expects specific parameters to be bound as parameters within the query and NOT as values within the query. Unfortunately, we cannot adapt this third-party JDBC-driver to support these parameter-bindung changes. The only solution to continue working with this third-party JDBC-driver is to guarantee that statements are generated with parameters!

Of course, it is obvious that these application failures triggered by this third-party JDBC-driver are caused by #492. However, it is also obvious that these application failures are implementetation specific for OUR applications and **DO NOT REFLECT A GENERAL ISSUE**.

**To summarize:**
 * Where is the performance impacted? => no performance impact
 * Incorrect query results? => nope
 * Database failures? => yes, a third-party JDBC-driver expects prepared statements with parameters; this leads to applications failures!

## 3. Why known workarounds are not sufficient to overcome these implications
> I'm not sure I understand this point. It is correct that with the #492 change, EclipseLink no longer will use parameter markers for literal values for DB2. However, it is not correct that EclipseLink will no longer use prepared statements with parameters.

That is completely true, it is possible to configure EclipseLink to use prepared statements with parameters. I mentioned three workarounds:
* add Parameter-Expressions to every query
* property in persistence.xml (`shouldBindLiterals=true`)
* WLP Custom Persistence Provider adding property `shouldBindLiterals=true`

However, the majority of our queries DO NOT set the parameters explicitly with `query.setParameter(...)`.

Of course, we could implement one of the mentioned workarounds to overcome implication (1) and (2). However, we operate multiple WLP-applications with each having multiple hundreds of `persistence.xml` and thousands of queries. The implementation of an adequate workarounds is only partially feasible and would be error-prone and cost-intense. Thus, we ask for a global solution!

## 4. How to implement a global solution
> One idea may be to use a JVM property. Currently, EclipseLink does not process JVM properties for persistence properties.

IMO this is not true, but maybe I am wrong...Let me try to explain:

EclipseLink loads config-properties within the class `EntityManagerFactoryProvider.java`. If you check the method  `getConfigPropertyLogDebug(...)` you will recognize that also the source of `SystemProperties` is accepted by EclipseLink, for details:
https://github.com/eclipse-ee4j/eclipselink/blob/2.7.6/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/EntityManagerFactoryProvider.java#L170-L190

The boolean property in the method `useSystemAsDefault` defines whether `SystemProperties` are accepted. By default, this is true!

However, to load config-properties from `SystemProperties` the method `getConfigPropertyAsStringLogDebug(...)` MUST explicitly be called for a specific `PersistenceUnitProperties`, for example:
* PersistenceUnitProperties.JDBC_BIND_PARAMETERS => https://github.com/eclipse-ee4j/eclipselink/blob/b7c997804f39187fb2907636d8403d694a318cfb/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/EntityManagerSetupImpl.java#L2786
* PersistenceUnitProperties.JDBC_FORCE_BIND_PARAMETERS => https://github.com/eclipse-ee4j/eclipselink/blob/b7c997804f39187fb2907636d8403d694a318cfb/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/EntityManagerSetupImpl.java#L2791

The same could be implemented for the property `PersistenceUnitProperties.TARGET_DATABASE_PROPERTIES`!

At least I was just able to set the property `PersistenceUnitProperties.JDBC_BIND_PARAMETERS` based on a `SystemProperty` defined in the `jvm.options` within a WLP-application-server.

**Comment by dazey3** [comment]
@hubmat00 Thank you for the detailed description of the concerns of #492! This is very useful information.

### 1. DB Performance: limitation of reusing prepared statements from the DB's statement cache
> The reusage of prepared statements speeds up query execution by decreasing the parsing and query planning overhead of each execution. However, this reusage of queries can only be achieved when binding literal values as parameters.

Reading this topic, and sub-topics, I do not see a requirement that only statements containing parameters will be cached: https://www.ibm.com/support/knowledgecenter/en/SSEPEK_11.0.0/perf/src/tpc/db2z_enabledynamicstmtcache.html

**Perhaps you can elaborate on your usage of JPA to create queries?** 

I think it is important to understand how you are using JPA to create your queries to understand why you are not seeing the benefit of prepared statement cache lookups. For instance, lets consider the following CriteriaBuilder example. Note that CriteriaBuilder is designed for programmatically creating JPQL queries, so CriteriaBuilder API calls can be translated to JPQL query calls:
```java
    int intValue = 100;
    ...
    Predicate intPredicate = criteriabuilder.equal(root.get(SimpleEntity_.intAttribute), intValue);
    criteriaquery.where(intPredicate);
    ...
    TypedQuery<SimpleEntity> query = em.createQuery(criteriaquery);
    List<SimpleEntity> result = query.getResultList();
```
this CriteriaBuilder roughly translates to the following simple JPQL example:
``` java
    int intValue = 100;
    String jpql = "SELECT e from SimpleEntity e WHERE e.intAttribute = " + intValue;
    TypedQuery<SimpleEntity> query = em.createQuery(jpql, SimpleEntity.class);
```
and both create the following SQL query, after #492:
```
[EL Fine]: sql: ... SELECT ID, INTATTRIBUTE FROM SIMPLEENTITY WHERE (INTATTRIBUTE = 100)
```

First, I verified that EclipseLink calls `java.sql.Connection.prepareStatement(String sql)` to create the PreparedStatement for both CriteriaBuilder and JPQL examples. I assume this is where the DB2 prepared statement cache would then be checked, but I am not a DB2 configuration expert and I will need to investigate this.
Second, this query will ALWAYS pass the value of "100". From my understanding, this query would be cached and found again when run in the future, yes? However, it would only be found if the EXACT same query is attempted to be prepared; with the value of "100". If the "intValue" variable changes value... then it wont be found in the cache and performance won't be very good. But if the "intValue" is changing dynamically, and you want the performance benefit of parameter binding, why not define "intValue" to be a parameter?:
```
    int intValue = 100;
    ...
    // Addition of ParameterExpression to declare the value as a query parameter
    ParameterExpression<Integer> intParameter = criteriabuilder.parameter(Integer.class);
    Predicate intPredicate = criteriabuilder.equal(root.get(SimpleEntity_.intAttribute), intParameter);
    criteriaquery.where(intPredicate);
    ...
    TypedQuery<SimpleEntity> query = em.createQuery(criteriaquery);
    // Setting the parameter value to be used
    query.setParameter(intParameter, intValue);
    List<SimpleEntity> result = query.getResultList();
```
which roughly translates to the following simple JPQL example:
``` java
    int intValue = 100;
    String jpql = "SELECT e from SimpleEntity e WHERE e.intAttribute = ?1";
    TypedQuery<SimpleEntity> query = em.createQuery(jpql, SimpleEntity.class);
    query.setParameter(1, intValue);
```
I feel it is important to point out this minor distinction because it was one of the reasons for #492 in the first place. If the application is going to pass a literal value, EclipseLink would treat that value as a literal. If the application would pass a parameter value, then EclipseLink would treat that value as a parameter. No more arbitrary, black box decisions concerning parameter binding. I would argue that the usage of parameters in the second part above is "correct" if the intention of the application is to mandate parameter binding.

Also, is it possible for you to elaborate on the noticed performance difference? I would be interested in hearing some numbers on before/after query times, numbers of queries being run, tests to verify. It would also be interesting to contrast with performance before #492, but with prepared statement cache disabled. I am not a DB2 expert, but I did find these instructions in which DB2 may already be saving performance information? (https://www.ibm.com/support/knowledgecenter/en/SSEPEK_11.0.0/perf/src/tpc/db2z_findinfostmtcache.html)

----

### 2. Application failures based on missing parameters

I find this very interesting and unexpected! Can you provide more details as to an example of this use case (query, driver name/version)? If you believe these failures are valid (not in violation of the JDBC spec or DB2 

I am aware of a couple examples in DB2 where it is illegal to use parameter markers (for instance, in the THEN/ELSE section of CASE expressions), but I am not aware of an example where it may be illegal to use literal values instead of parameter markers.

----

### 3. Why known workarounds are not sufficient to overcome these implications
> it is possible to configure EclipseLink to use prepared statements with parameters. I mentioned three workarounds:
> * [A] add Parameter-Expressions to every query
> * [B] property in persistence.xml (`shouldBindLiterals=true`)
> * [C] WLP Custom Persistence Provider adding property `shouldBindLiterals=true`
*Note: I added letters to more easily differentiate in discussion

I agree with these "workarounds". I would argue though that option `A` is the "correct" way to implement your JPQL query if you want parameter binding/performance benefits and is not necessarily a "workaround". But I already discussed that in section 1 above.

> we could implement one of the mentioned workarounds to overcome implication (1) and (2). However, we operate multiple WLP-applications with each having multiple hundreds of persistence.xml and thousands of queries. The implementation of an adequate workarounds is only partially feasible and would be error-prone and cost-intense. Thus, we ask for a global solution!

I understand that a large code fix to your application is not ideal; especially for a performance improvement like this. A global solution, such as a JVM/System property would be best as a work around in lieu of correcting the application logic.
----

### 4. How to implement a global solution
> > One idea may be to use a JVM property. Currently, EclipseLink does not process JVM properties for persistence properties.
> IMO this is not true

You are correct here. I double checked and it seems that EclipseLink's use of `org.eclipse.persistence.internal.jpa.EntityManagerFactoryProvider.getConfigPropertyLogDebug` is a bit... inconsistent. EclipseLink does not currently process the "eclipselink.target-database-properties" property as a System property. There are also several other EclipseLink provider properties that also are not read as System properties. I can't think of a reason off the top of my head NOT to expand the utility to include "eclipselink.target-database-properties" to process as a System property, but I'll have to investigate a bit more to double check.

> At least I was just able to set the property PersistenceUnitProperties.JDBC_BIND_PARAMETERS based on a SystemProperty defined in the jvm.options within a WLP-application-server.

This would be the correct way to set a JVM property for WebSphere. I'll look into improving EclipseLink to process this property as a System property rather than ONLY as a persistence unit property.

**Comment by hubmat00** [comment]
@dazey3 I think now I can relate to your unclarities on this matter and try my best to clear them up.

### 1. DB Performance: limitation of reusing prepared statements from the DB's statement cache
> I think it is important to understand how you are using JPA to create your queries to understand why you are not seeing the benefit of prepared statement cache lookups.

Yes, this is the solution to nail down our concerns about db performance.

I target our concern regarding db performance to our existing JPA-queries mostly based on CriteriaBuilder and JPQL, in which after #492 our sql-queries definitely change. The reason why our current sql-queries change with #492 is that so far we DO NOT explicitly set parameters with:
```
// Setting the parameter value to be used
query.setParameter(intParameter, intValue);
```
**We count on the (your mentioned) black box decisions of EclipseLink's parameter binding that our queries bind literals as parameters**. With #492 this black box decision regarding parameter binding definitely changes and, thus, our sql-queries change accordingly (literals get no longer bound as parameters). The combination of `literals get no longer bound as parameters` and `we don't explicitly set parameters` leads to our misery!

And you perfectly pointed out why #492 is important:
> If the application is going to pass a literal value, EclipseLink would treat that value as a literal. If the application would pass a parameter value, then EclipseLink would treat that value as a parameter.

I totally agree with you:
>No more arbitrary, black box decisions concerning parameter binding. I would argue that the usage of parameters in the second part above is "correct" if the intention of the application is to mandate parameter binding.

**For (our) current applications this is exaclty the source of concerns regarding db performance:**
The combination of `literals get no longer bound as parameters` and `don't explicitly setting literal parameters for queries based on CriteriaBuilder and JPQL` leads to changed sql-queries which in turn decreases the db performance. The decrease of db performance comes from #492 where JPA-queries fabricate cache misses, whereas without #492 the same JPA-queries would have been triggered a cache hit.

**I am not a db expert, but I think the following statements about #492 can be derived from the discussion above (please correct if I am wrong):**
To summarize, the changes of #492 result in a decreased backward-compatibility of JPA-queries that `don't explicitly set literal parameters` by changing the generated sql-output of EclipseLink`s statement generator. These changes of #492 potentially lead to a decreased db performance by not optimally using db's statement-cache because more specific-entries are stored in the statement cache.

From now on for new developments or refactorings we will definitely set literal parameters explicitly! However, for the high number of queries in (our) existing applications we are not able to refactor all queries at once as it would be too cumbersome and error-prone.

### 2. Application failures based on missing parameters
> I find this very interesting and unexpected! Can you provide more details as to an example of this use case (query, driver name/version)? If you believe these failures are valid (not in violation of the JDBC spec or DB2

This mentioned custom JDBC driver wraps the official DB2-JDBC-driver. This custom driver is mainly used to solve staging, accounting, tracing and multi-tenancy transparently for WLP-applications. And yes, this driver is build on the assumption that `literals get bound as parameters`. With #492 this assumption is no longer fulfilled which in turn leads to our mentioned application failures.

Actually I am not sure whether this custom driver is spec conform. A fair guess, it is NOT. However, it doesn't really matter, we can't refactor/modify this custom-made-driver, thus, we are limited to a solution outside the scope of this driver. The most efficient would be to globally set `shouldBindLiterals=true`.

One remark: for new applications we already eliminated this third-party JDBC driver by using JPA-spec-conform solutions.

### 3. Why known workarounds are not sufficient to overcome these implications
I agree with your input, nothing to add!

### 4. How to implement a global solution
> This would be the correct way to set a JVM property for WebSphere. I'll look into improving EclipseLink to process this property as a System property rather than ONLY as a persistence unit property.

We are really looking forward to your implementation to support `shouldBindLiterals=true` within upcoming versions of WLP. Until then I really thank you for your help and professional collaboration!
